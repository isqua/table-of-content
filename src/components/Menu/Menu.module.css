.menu {
    font-size: 13px;
    font-style: normal;
    font-weight: 400;
    line-height: 20px;

    --menu-item-block-padding: 8px;
    --menu-item-inline-padding: 22px;
    --menu-item-level-padding: 16px;
}

.list {
    display: flex;
    flex-direction: column;
    padding: 0;
    margin: 0;
    list-style-type: none;
}
