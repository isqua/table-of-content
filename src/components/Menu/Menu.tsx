import { useLocation } from 'react-router-dom'

import { type TableOfContent } from '../../features/toc'
import { Section } from './Section/Section'
import { MenuProvider } from './Context/MenuProvider'

import styles from './Menu.module.css'

type MenuProps = {
    toc: TableOfContent
}

export function Menu({ toc }: MenuProps): JSX.Element {
    const location = useLocation()
    const currentUrl = location.pathname.replace(/^\//, '')

    return (
        <nav className={styles.menu}>
            <ul className={styles.list}>
                <MenuProvider toc={toc} url={currentUrl}>
                    <Section parentId='' level={0} />
                </MenuProvider>
            </ul>
        </nav>
    )
}
