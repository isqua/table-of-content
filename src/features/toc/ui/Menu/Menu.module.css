.menu {
    font-size: 13px;
    font-style: normal;
    font-weight: 400;
    line-height: 20px;

    --menu-item-block-padding: 8px;
    --menu-item-inline-start-padding: 26px;
    --menu-item-inline-end-padding: 22px;
    --menu-item-level-padding: 16px;

    /* Align chevrons with text even for multiline rows */
    --menu-chevron-vertical-correction: 2px;
    --menu-chevron-horizontal-correction: -20px;
}
