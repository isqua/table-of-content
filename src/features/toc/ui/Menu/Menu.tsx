
import type { TableOfContent } from '../../types'
import { MenuProvider } from './Context/MenuProvider'
import { Filter } from './Filter/Filter'
import { List } from './List/List'

import styles from './Menu.module.css'

export type MenuProps = {
    toc: TableOfContent
    activeUrl: string
    isLoading?: boolean
}

export function Menu({ toc, activeUrl, isLoading }: MenuProps): JSX.Element {
    return (
        <nav className={styles.menu}>
            <MenuProvider toc={toc} url={activeUrl} isLoading={isLoading}>
                <Filter />
                <List />
            </MenuProvider>
        </nav>
    )
}
