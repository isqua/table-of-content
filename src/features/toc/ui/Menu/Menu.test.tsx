import { act, fireEvent, screen } from '@testing-library/react'
import type { PropsWithChildren } from 'react'
import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest'

import { renderInApp } from '../../../../test'
import tocFlat from '../../../../test/fixtures/toc/flat.json'
import tocTwoLevels from '../../../../test/fixtures/toc/two-levels.json'
import type { TableOfContent } from '../../types'
import { Menu, MenuProps } from './Menu'

const spinnerText = 'Loading'

const getSpinner = () => screen.getByLabelText(spinnerText)

const updateFilter = (value: string) => {
    fireEvent.change(screen.getByRole('textbox'), {
        target: { value },
    })
}

const renderMenu = (props: MenuProps, currentUrl?: string) => {
    renderInApp(<Menu {...props} />, { url: currentUrl })
}

const assertMenuSnapshot = () => {
    expect(screen.getByRole('navigation')).toMatchSnapshot()
}

vi.mock('react-transition-group', () => {
    const FakeTransitionGroup = vi.fn(
        ({ children }: PropsWithChildren) => children
    )

    const FakeTransition = vi.fn(
        ({ children }: PropsWithChildren) => children
    )

    const FakeCSSTransition = vi.fn(
        (props: PropsWithChildren<{ in: boolean }>) => props.in ?
            <FakeTransition>{props.children}</FakeTransition> :
            null,
    )

    return {
        TransitionGroup: FakeTransitionGroup,
        CSSTransition: FakeCSSTransition,
        Transition: FakeTransition,
    }
})

describe('features/toc/ui/Menu', () => {
    describe('loading', () => {
        it('should render skeletons while the TOC is loading', () => {
            const toc: TableOfContent = tocTwoLevels
            const currentUrl = '/bar.html'

            renderMenu({ toc, isLoading: true }, currentUrl)

            assertMenuSnapshot()
        })
    })

    describe('render', () => {
        it('should build a menu and highlight the current page', () => {
            const toc: TableOfContent = tocFlat
            const currentUrl = '/bar.html'

            renderMenu({ toc }, currentUrl)

            assertMenuSnapshot()
        })

        it('should build a two-levels menu and open all parents containing the current page', () => {
            const toc: TableOfContent = tocTwoLevels
            const currentUrl = '/bar-install.html'

            renderMenu({ toc }, currentUrl)

            assertMenuSnapshot()
        })
    })

    describe('toggling', () => {
        it('should close a submenu when clicking on a chevron', () => {
            const toc: TableOfContent = tocTwoLevels
            const currentUrl = '/bar-install.html'

            renderMenu({ toc }, currentUrl)

            expect(screen.getByRole('navigation')).toMatchSnapshot()

            fireEvent.click(screen.getByRole('button', { expanded: true }))

            assertMenuSnapshot()
        })

        it('should open a submenu when clicking on an item with children', () => {
            const toc: TableOfContent = tocTwoLevels
            const currentUrl = '/foo.html'

            renderMenu({ toc }, currentUrl)

            fireEvent.click(screen.getByRole('link', { name: 'Bar' }))

            expect(screen.getByRole('button', { expanded: true })).toBeInTheDocument()

            expect(screen.getByText('Bar: Install')).toBeInTheDocument()
        })
    })

    describe('filtering', () => {
        beforeAll(() => {
            vi.useFakeTimers()
        })

        afterAll(() => {
            vi.useRealTimers()
        })

        it('should show a loader when filtering', () => {
            const toc: TableOfContent = tocTwoLevels
            const currentUrl = ''
            const searchText = 'bar'

            renderMenu({ toc }, currentUrl)

            updateFilter(searchText)

            expect(getSpinner()).toBeInTheDocument()

            act(() => {
                vi.runAllTimers()
            })
        })

        it('should filter items by text', () => {
            const toc: TableOfContent = tocTwoLevels
            const currentUrl = ''
            const searchText = 'bar'

            renderMenu({ toc }, currentUrl)

            updateFilter(searchText)

            act(() => {
                vi.runAllTimers()
            })

            expect(screen.queryByLabelText(spinnerText)).not.toBeInTheDocument()

            assertMenuSnapshot()
        })
    })
})
