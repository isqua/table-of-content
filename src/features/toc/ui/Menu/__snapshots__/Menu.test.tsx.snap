// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`features/toc/ui/Menu > filtering > should filter items by text 1`] = `
<nav
  class="_menu_f33640"
>
  <div
    class="_filter_daddd4"
  >
    <div
      class="_input_98f327"
    >
      <input
        class="_control_98f327"
        placeholder="Filter menu"
        type="text"
      />
    </div>
  </div>
  <ul
    class="_list_ae4333"
  >
    <li
      aria-level="1"
      class="_item_c21f54"
    >
      <a
        class="_link_c21f54"
        href="/bar.html"
      >
        <span
          class="_text_c21f54"
        >
          <svg
            aria-expanded="true"
            class="_chevron_eafbf3 _toggle_c21f54 _open_eafbf3"
            fill="none"
            height="16"
            role="button"
            width="16"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M6 13 11 8 6 3v9Z"
              fill="currentColor"
            />
          </svg>
          Bar
        </span>
      </a>
    </li>
    <li
      aria-level="2"
      class="_item_c21f54"
    >
      <a
        class="_link_c21f54"
        href="/bar-install.html"
      >
        <span
          class="_text_c21f54"
        >
          Bar: Install
        </span>
      </a>
    </li>
    <li
      aria-level="2"
      class="_item_c21f54"
    >
      <a
        class="_link_c21f54"
        href="/bar-features.html"
      >
        <span
          class="_text_c21f54"
        >
          Bar: Features
        </span>
      </a>
    </li>
  </ul>
</nav>
`;

exports[`features/toc/ui/Menu > loading > should render skeletons while the TOC is loading 1`] = `
<nav
  class="_menu_f33640"
>
  <ul
    class="_list_ae4333"
  >
    <li
      aria-level="1"
      class="_item_c21f54"
    >
      <span
        class="_link_c21f54 _disabled_c21f54"
      >
        <span
          class="_text_c21f54"
        >
          <span
            class="_skeleton_2cc9eb _loader_c21f54"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="2"
      class="_item_c21f54"
    >
      <span
        class="_link_c21f54 _disabled_c21f54"
      >
        <span
          class="_text_c21f54"
        >
          <span
            class="_skeleton_2cc9eb _loader_c21f54"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="2"
      class="_item_c21f54"
    >
      <span
        class="_link_c21f54 _disabled_c21f54"
      >
        <span
          class="_text_c21f54"
        >
          <span
            class="_skeleton_2cc9eb _loader_c21f54"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="1"
      class="_item_c21f54"
    >
      <span
        class="_link_c21f54 _disabled_c21f54"
      >
        <span
          class="_text_c21f54"
        >
          <span
            class="_skeleton_2cc9eb _loader_c21f54"
          />
        </span>
      </span>
    </li>
    <li
      aria-level="1"
      class="_item_c21f54"
    >
      <span
        class="_link_c21f54 _disabled_c21f54"
      >
        <span
          class="_text_c21f54"
        >
          <span
            class="_skeleton_2cc9eb _loader_c21f54"
          />
        </span>
      </span>
    </li>
  </ul>
</nav>
`;

exports[`features/toc/ui/Menu > render > should build a menu and highlight the current page 1`] = `
<nav
  class="_menu_f33640"
>
  <div
    class="_filter_daddd4"
  >
    <div
      class="_input_98f327"
    >
      <input
        class="_control_98f327"
        placeholder="Filter menu"
        type="text"
      />
    </div>
  </div>
  <ul
    class="_list_ae4333"
  >
    <li
      aria-level="1"
      class="_item_c21f54"
    >
      <a
        class="_link_c21f54 _active_c21f54"
        href="/bar.html"
      >
        <span
          class="_text_c21f54"
        >
          Bar
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_c21f54"
    >
      <a
        class="_link_c21f54"
        href="/baz.html"
      >
        <span
          class="_text_c21f54"
        >
          Baz
        </span>
      </a>
    </li>
    <li
      aria-level="1"
      class="_item_c21f54"
    >
      <a
        class="_link_c21f54"
        href="/foo.html"
      >
        <span
          class="_text_c21f54"
        >
          Foo
        </span>
      </a>
    </li>
  </ul>
</nav>
`;
